onnxruntime/core/optimizer/propagate_cast_ops.cc

```diff
@@ -0,0 +1,709 @@
+            // convert initializers if any from float to float16.
+            if (input_def != node_arg) {
+              if (IsType(*input_def, TensorProto_DataType_FLOAT)) {
+                require_cast.insert(node_arg);
+                return;
+              }
+            }
+          }
+          for (NodeArg* node_output : node->MutableOutputDefs()) {
+            if (IsType(*node_output, TensorProto_DataType_FLOAT) &&
+                require_cast.find(node_output) == require_cast.end() &&
+                require_type_change.find(node_output) == require_type_change.end()) {
+              SearchDownstream(graph, node_output, require_cast, require_type_change, removed_nodes, level);
+              if (require_cast.find(node_output) == require_cast.end()) {
+                require_type_change.insert(node_output);
+              }
+            }
+          }
+        }
+      }
+    }
+  }
+}
+
+// ConcatNames
+// Collects all the names from the pointers of the objects stores in the container class C
+// the class should have a member functions returning a string (or a ref).
+template <typename C, typename T = typename C::value_type>
+static std::string ConcatNames(C const& items) {
+  std::vector<std::string> names;
+  std::transform(items.begin(), items.end(), back_inserter(names), [](T n) { return n->Name(); });
+  return std::accumulate(names.begin(), names.end(), std::string(), [](const std::string& a, const std::string& b) { return a + ", " + b; });
+}
+
+// Change the elem_type of the given NodeArgs from FLOAT to FLOAT16.
+static void ChangeTypeToFP16(std::unordered_set<NodeArg*>& require_type_change, const logging::Logger& logger) {
+  ONNX_NAMESPACE::TypeProto type_proto;
+  type_proto.mutable_tensor_type()->set_elem_type(TensorProto::FLOAT16);
+  for (NodeArg* input : require_type_change) {
+    if (IsType(*input, TensorProto::FLOAT)) {
+      input->UpdateTypeAndShape(type_proto, true, true, logger);
+    }
+  }
+}
+
+// PropagateForwards
+// Propagate FP32 Cast operations forwards (downstream)
+// Using SearchDownStream search the graph for Cast FP16 safe/allowed operations to expand
+// the float16 computation region.
+// The required_cast vector is the collection of nodes that require float cast.
+// All nodeargs on a path down to any of the
+// frontier nodes require type change from FLOAT to FLOAT16.
+// require_type_change consists of such nodes.  All the frontier nodes require fp32 cast
+// The input node is expected to be non-nullptr
+static bool PropagateForwards(Graph& graph, Node* node,
+                              std::deque<onnxruntime::NodeIndex>& removed_nodes,
+                              size_t level,
+                              const logging::Logger& logger) {
+  ORT_ENFORCE(node != nullptr);
+  bool modified = false;
+  std::unordered_set<NodeArg*> require_cast;
+  std::unordered_set<NodeArg*> require_type_change;
+  NodeArg* cast_output = node->MutableOutputDefs()[0];
```

Comment:
>MutableOutputDefs [](start = 31, length = 17)

This is non-block comment. 

Alternatively, we can use 
const NodeArg* cast_output = node->OutputDefs()[0];

Reply:
Find the answer below.